packages/slate-editor/src/extensions/galleries/components/GalleryElement.tsx

```diff
@@ -1,36 +1,133 @@
 import type { GalleryNode } from '@prezly/slate-types';
-import { UploadcareImage } from '@prezly/uploadcare';
-import React from 'react';
+import { awaitUploads, UPLOADCARE_FILE_DATA_KEY, UploadcareImage } from '@prezly/uploadcare';
+import { noop } from '@technically/lodash';
+import React, { useState } from 'react';
 import type { Editor } from 'slate';
+import { Transforms } from 'slate';
 import type { RenderElementProps } from 'slate-react';
+import { useSlateStatic } from 'slate-react';
 
 import { EditorBlock } from '#components';
-import { useSize } from '#lib';
+import { useLatest, useSize } from '#lib';
+
+import type { MediaGalleryOptions } from '#modules/uploadcare';
+import { UploadcareEditor } from '#modules/uploadcare';
+
+import { shuffleImages } from '../lib';
+import { updateGallery } from '../transforms';
 
 import { Gallery } from './Gallery';
 import { GalleryMenu } from './GalleryMenu';
 
 interface Props extends RenderElementProps {
     availableWidth: number;
     element: GalleryNode;
-    onEdit: (editor: Editor) => void;
-    withWidthOption: boolean;
+    onEdit?: (editor: Editor, gallery: GalleryNode) => void;
+    onEdited?: (
+        editor: Editor,
+        gallery: GalleryNode,
+        extra: {
+            successfulUploads: number;
+            failedUploads: Error[];
+        },
+    ) => void;
+    onShuffled?: (editor: Editor, updated: GalleryNode, original: GalleryNode) => void;
+    withMediaGalleryTab: false | { enabled: true; newsroom: NewsroomRef };
+    withLayoutOptions: boolean;
 }
 
 export function GalleryElement({
     availableWidth,
     attributes,
     children,
     element,
-    onEdit,
-    withWidthOption,
+    onEdit = noop,
+    onEdited = noop,
+    onShuffled = noop,
+    withMediaGalleryTab,
+    withLayoutOptions,
 }: Props) {
+    const editor = useSlateStatic();
     const [sizer, size] = useSize(Sizer, { width: availableWidth });
+    const [isUploading, setUploading] = useState(false);
+    const callbacks = useLatest({ onEdit, onEdited, onShuffled });
+
+    async function handleEdit() {
+        callbacks.current.onEdit(editor, element);
+
+        const files = element.images.map(({ caption, file }) => {
+            const uploadcareImage = UploadcareImage.createFromPrezlyStoragePayload(file);
+            uploadcareImage[UPLOADCARE_FILE_DATA_KEY] = { caption };
+            return uploadcareImage;
+        });
+
+        async function upload() {
+            const filePromises = await UploadcareEditor.upload(editor, {
+                ...getMediaGalleryParameters(withMediaGalleryTab),
+                captions: true,
+                files,
+                imagesOnly: true,
+                multiple: true,
+            });
+
+            if (!filePromises) {
+                return undefined;
+            }
+
+            return awaitUploads(filePromises);
+        }
+
+        setUploading(true);
+
+        try {
+            const result = await upload();
+
```

Comment:
Shouldn't `setUploading(true)` live in the `try` block?

Reply:
Why? It cannot fail AFAIK